nav2_planner/src/planner_server.cpp

```diff
@@ -510,7 +516,52 @@ PlannerServer::publishPlan(const nav_msgs::msg::Path & path)
   }
 }
 
-}  // namespace nav2_planner
+void PlannerServer::isPathValid(
+  const std::shared_ptr<nav2_msgs::srv::IsPathValid::Request> request,
+  std::shared_ptr<nav2_msgs::srv::IsPathValid::Response> response)
+{
+  unsigned int mx = 0;
+  unsigned int my = 0;
+
+  response->is_valid = true;
+
+  if (request->path.poses.empty() ) {
+    response->is_valid = false;
+  } else {
+    geometry_msgs::msg::PoseStamped current_pose;
+    unsigned int closest_pose_index = 0;
+    if (costmap_ros_->getRobotPose(current_pose)) {
+      tf2::Vector3 curr_position(current_pose.pose.position.x,
+        current_pose.pose.position.y,
+        current_pose.pose.position.z);
+      // Find index of pose closest to the robot
+      float closest_distance = 0;
+      for (unsigned int i = 0; i < request->path.poses.size(); ++i) {
```

Comment:
To make this a little more robust, may I suggest that you also track the last distance(s) as we march through the path, it should be getting closer and then further from the robot (or if its the start, just further). So if we see the trend that its getting further, we can stop checking and exit.

That helps with the case that there are loops in the path and the incorrect point is selected because it is floating-point error closer to the robot than the actual point of interest (which is a problem right now for the controllers that use a similar technique as this)

Reply:
Done